addons/sale/models/sale.py

```diff
@@ -201,6 +201,13 @@ def _get_payment_type(self):
         ('confirmation_date_conditional_required', "CHECK( (state IN ('sale', 'done') AND confirmation_date IS NOT NULL) OR state NOT IN ('sale', 'done') )", "A confirmed sales order requires a confirmation date."),
     ]
 
+    @api.constrains('company_id', 'order_line')
+    def _check_order_line_company_id(self):
+        for order in self:
+            companies = order.mapped('order_line.product_id.company_id')
+            if companies and companies != order.company_id:
+                raise ValidationError((_("Your quotation contains products from company %s whereas your quotation belongs to company %s. \n Please change the company of your quotation or remove the products from other companies.") % (companies.mapped('display_name'), order.company_id.display_name)))
+
```

Comment:
same comment as above regarding listing the products in the error message and the mapped calls